azerothcore/azerothcore-wotlk
# feat(Core/Warden): Allow sending of custom lua payloads through Warden.

## Changes Proposed:
-  Allow modules to access warden from the player session and send custom lua payloads.

## Issues Addressed:

- Closes $1700 bounty which can be paid to me in small 1 cent denominations.

## SOURCE:


## Tests Performed:

- Ensured sure normal Warden payloads are sent with the custom payload.
- Ensured if Warden is interrupted through ForceChecks that the result is not counted raising a false positive.
- Ensured Warden functions normally without modules.


## How to Test the Changes:


1. Clone my [wardenpr branch](https://github.com/AnchyDev/azerothcore-wotlk/tree/wardenpr).
2. Clone [my test module](https://github.com/AnchyDev/WardenLuaTestMod) or clone my [Breaking News Override](https://github.com/AnchyDev/BreakingNewsOverride/tree/wardenpr) module (wardenpr branch).
3. Enable the test module in the config.
4. Enable warden debug logging in world config.
`#Logger.warden=5,Console Server` -> `Logger.warden=5,Console Server`
`Appender.Console=1,4,0,"1 9 3 6 5 8"` -> `Appender.Console=1,5,0,"1 9 3 6 5 8"`
5. Observe that all 5 unit tests pass in the world server console, and when you type in chat it echoes back your message. (If you are testing breaking news you should see it on the character screen.)

## Known Issues and TODO List:


- If you are sending constant forced payloads, Warden will never check results making it basically useless. It's a good idea to only use this to setup an addon message listener on the client then stop using Warden payloads entirely. 


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
